Hi Wink!

> With gcc on linux I to have move the '$!<' before $!Libs so main.o is before the libraries:

Oh right, that's quite possibly necessary. I sometimes forget about that. I was testing on Mac with Clang, and it works there.

> Then if I execute 'ninja -v -j14' I see the failure that libL001.a, libL003.a and libL004.a haven't been created yet. And very oddly step [21/22] is the link step for testapp while step[22/22] is creating archive for libL003.a, this definitely feels wrong.

You're right, there must be something wrong. It should only link after all libraries have been built, since the testapp depends on all libraries. Unfortunately, I can't reproduce this on Mac. I will check it on Linux soon.

> Another thing, what I'm looking for is not so much how long it takes for cmake, meson or creator to run the first time, but more importantly how long it takes to just do the build steps there after.

I understand. But I think that highly depends on the back-end, for which it is Ninja in all three cases (except you can also choose other output formats for CMake). For Creator, you can use the `-e` option to force an export only and not invoke Ninja under the hood.

> In your case you're hiding ninja behind creator so your code is part of the build step, but that's not the case with cmake or meson.

I don't necessarily _hide_ it, it's always there and you can use Ninja directly to run the build as well after Creator has generated it. But I found it to be more convenient to use the Creator command-line interface over Ninja. I understand that for projects of this size, that is hardly an option due to the long startup time.

---

**Update** I can't reproduce it on Linux either. It builds everything in order. What Ninja version are you using? The Ninja version on my Linux machine is rather old (1.3.4).

On another side note, I'm wondering why GCC complains about the undefined references? Does GCC not accept absolute paths to static libraries? I used `nm` to check if the `libLXXX.a` files really contain the functions, and they do.

```
niklas@vbox-ubuntu:~/Desktop/gen_srcs$ ninja -j 14
[22/22] gcc -std=c99 /home/niklas/Desktop/gen_srcs/test/buil...in.o -o /home/niklas/Desktop/gen_srcs/test/build/bin/testapp
FAILED: gcc -std=c99 /home/niklas/Desktop/gen_srcs/test/build/libs/libL000.a /home/niklas/Desktop/gen_srcs/test/build/libs/libL001.a /home/niklas/Desktop/gen_srcs/test/build/libs/libL002.a /home/niklas/Desktop/gen_srcs/test/build/libs/libL003.a /home/niklas/Desktop/gen_srcs/test/build/libs/libL004.a /home/niklas/Desktop/gen_srcs/test/build/libs/libL005.a /home/niklas/Desktop/gen_srcs/test/build/libs/libL006.a /home/niklas/Desktop/gen_srcs/test/build/libs/libL007.a /home/niklas/Desktop/gen_srcs/test/build/libs/libL008.a /home/niklas/Desktop/gen_srcs/test/build/libs/libL009.a /home/niklas/Desktop/gen_srcs/test/build/obj/testapp/main.o -o /home/niklas/Desktop/gen_srcs/test/build/bin/testapp
/home/niklas/Desktop/gen_srcs/test/build/obj/testapp/main.o: In function `main':
main.c:(.text+0x5): undefined reference to `func1'
main.c:(.text+0xa): undefined reference to `func2'
main.c:(.text+0xf): undefined reference to `func3'
main.c:(.text+0x14): undefined reference to `func4'
main.c:(.text+0x19): undefined reference to `func5'
main.c:(.text+0x1e): undefined reference to `func6'
main.c:(.text+0x23): undefined reference to `func7'
main.c:(.text+0x28): undefined reference to `func8'
main.c:(.text+0x2d): undefined reference to `func9'
main.c:(.text+0x32): undefined reference to `func10'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```

**Update** Nevermind, I forgot to re-generate the project files with `gen_srcs.py` after moving `$!Libs` after `$!<`.

--- a/gen_srcs.py
+++ b/gen_srcs.py
@@ -559,7 +559,7 @@ def obj():
 
 @target(abstract=True)
 def bin():
-  bin.build('$Objects', '$Bin', '$c:cc $CFlags $!Libs $!< $(c:binout $@)')
+  bin.build('$Objects', '$Bin', '$c:cc $CFlags $!< $!Libs $(c:binout $@)')
 '''.strip())
 
         with open(os.path.join(root_path, 'template.lib.creator'), 'w') as fp:

--- a/gen_srcs.py
+++ b/gen_srcs.py
@@ -1,7 +1,7 @@
 #!/usr/bin/env python3
 # Generate C source files
 
-import os.path, sys, argparse, textwrap
+import os.path, sys, argparse
 
 version = '0.0.1'
 
@@ -606,12 +606,15 @@ def lib():
 
             libs = ''
             includes = ''
+            deps = []
             for lib in app.getLibraries():
                 fp.write("load('{0}')\n".format(lib.getLibName()))
+                deps.append(lib.getLibName() + ':lib')
                 libs += ';${0}:Lib'.format(lib.getLibName())
                 includes += ';${0}:Includes'.format(lib.getLibName())
             fp.write("append('Libs', {0!r})\n".format(libs))
             fp.write("append('Includes', {0!r})\n".format(includes))
+            fp.write("[bin.requires(x) for x in %r]\n" % deps)
 
     def begLibBuilder(self, libraries_path):
         self._libs_file = open(os.path.join(libraries_path, '.creator'), 'w')

